engine/api/workflow/dao_node_run.go

```diff
@@ -425,6 +414,28 @@ func GetNodeRunBuildCommits(db gorp.SqlExecutor, store cache.Store, p *sdk.Proje
 	return res, cur, nil
 }
 
+// PreviousNodeRun find previous node run
+func PreviousNodeRun(db gorp.SqlExecutor, nr sdk.WorkflowNodeRun, n sdk.WorkflowNode, workflowID int64) (sdk.WorkflowNodeRun, error) {
+	query := `
+					SELECT workflow_node_run.* FROM workflow_node_run
+					JOIN workflow_node ON workflow_node.name = $1 AND workflow_node.workflow_id = $2
+					WHERE vcs_branch = $3 AND workflow_node_run.num <= $4 AND workflow_node_run.workflow_node_id = $5 AND workflow_node_run.id != $6
+					ORDER BY workflow_node_run.num, workflow_node_run.sub_num DESC
+					LIMIT 1 
+				`
+	var nodeRun sdk.WorkflowNodeRun
+	var rr = NodeRun{}
+	if err := db.SelectOne(&rr, query, n.Name, workflowID, nr.VCSBranch, nr.Number, nr.WorkflowNodeID, nr.ID); err != nil {
+		return nodeRun, sdk.WrapError(err, "PreviousNodeRun> Cannot load previous RUN: %s [%s %d %s %d]", query, n.Name, workflowID, nr.VCSBranch, nr.Number, nr.WorkflowNodeID)
+	}
+	pNodeRun, errF := fromDBNodeRun(rr)
+	if errF != nil {
+		return nodeRun, sdk.WrapError(errF, "PreviousNodeRun> Cannot read node run")
+	}
+	nodeRun = *pNodeRun
```

Comment:
sure that pNodeRun won't be nil event if there isn't error ?

Reply:
yes